### ISSUE TITLE
ReQL .map creates false variable name

### ISSUE THREAD

**Issue by Xrave** [rethinkdb/rethinkdb#3959]
Hey, I found this behavior in ReQL and believe it is a bug. 
In any dataexplorer: (tested on both 1.15 and 1.16.x)
```
r.expr([{"a":1}, {"a":3}, {"a":4}]).do(function(subdoc){
    var c1;
    subdoc.map(function(itm){
      c1 = itm("a");
      return itm;
    });
    return c1;
});
```
expected output: `4`, or at least something that's not an error.
received output:
```
RqlCompileError: Variable name not found in:
r([{a: 1}, {a: 3}, {a: 4}]).do(function(var_25) { return var_26("a"); })
                                                         ^^^^^^         
```

I'm writing an Array-of-Object --> Object-of-Array conversion in ReQL and this is the core issue that I am hitting. [Here's the StackOverflow](http://stackoverflow.com/questions/29177094/rethinkdb-array-of-object-to-object-of-array) I posted if anyone is kind enough to help me out.

**Comment by mlucy** [comment]
Hi @Xrave,

The function you pass to `do` is actually called once to generate a sub-query which is then sent to the server, so you can't put complicated JS logic in there.  `subdoc.map` doesn't run any code; it just generates a sub-query, so when you return `c1` you aren't returning a value that was set while executing something, you're returning a sub-query (which happens to contain a variable which was introduced in the `map`, but that introduction was thrown away because you didn't return the `map`).  That's where the `var_26` comes from.

(If that wasn't clear, you can read more about it at http://rethinkdb.com/blog/lambda-functions/ .)

Basically, if you're doing variable assignment inside a ReQL function then what you're doing is probably incorrect.

Closing as not-a-bug, but feel free to ask more questions here.

**Comment by Xrave** [comment]
Thanks for the prompt response, @mlucy 
That makes sense and I sort of expected that to be happening since it goes away if I return the map directly. 

But what kind of query statements should I be making if I want to translate [{"a":1},{"a":2}] to {"a":[1,2]} on the fly? Would r.js() be a better choice for me then?

**Comment by danielmewes** [comment]
@Xrave you can use `concatMap` for that:
```js
r.expr({a:
  r.expr([{"a":1},{"a":2}]).concatMap([r.row('a')])
})
```

**Comment by mlucy** [comment]
@danielmewes -- I think you can just use a normal `map` there.

**Comment by Xrave** [comment]
@danielmewes Sorry, I simplified it for the purpose of demonstration, but there are unknown number of keys in each, like so: `[{"a":1, "b":2}, {"a":3}, {"c":4}]` and I want to transform all of them.

**Comment by Xrave** [comment]
Aha, fixed it. this seems to work:
```
r.expr([{"a":1, "b":2}, {"a":3}, {"c":4}]).do( r.js("(function(input){ var keys=['a','b','c'], output={}; input.forEach(function(e){ for(var p in e){ output[p] = (output[p] || []).concat(e[p]);  } }); return output; })"))
```
more readably
```
r.expr([{"a":1, "b":2}, {"a":3}, {"c":4}]).do( r.js("(function(input){ 
       var keys=['a','b','c'], output={}; 
       input.forEach(function(e){ 
            for(var p in e){ 
                  output[p] = (output[p] || []).concat(e[p]);  
       } }); 
       return output; })"))
```

**Comment by danielmewes** [comment]
Ah I see. Here's an alternative without `r.js`, though I wouldn't call it particularly simple:
```js
input = r.expr([{"a":1, "b":2}, {"a":3}, {"c":4}]);
keys = input.map(r.row.keys()).reduce(function(l, r) {return l.setUnion(r);});

keys.do(function (ks) {
  return ks.map(function(key) { return [key, input(key)] }).coerceTo('OBJECT');
})
```

This might become easier with https://github.com/rethinkdb/rethinkdb/issues/873

**Comment by danielmewes** [comment]
For email readers: I simplified the code.

**Comment by deontologician** [comment]
I put an explanation of @danielmewes solution on the [stackoverflow question](http://stackoverflow.com/a/29178265/180718) if anyone is interested. There were some good tricks there.

**Comment by mlucy** [comment]
Here's another bad solution:

```js
r.expr([{"a":1, "b":2}, {"a":3}, {"c":4}]).reduce(function (a, b) {
  return b.keys().map(function(key) {
    return [key, a(key).default(0).add(b(key))];
  }).coerceTo('object').do(function (rhs) {
    return a.merge(rhs);
  });
})
```

I agree #873 is probably the right way to make this easy.

**Comment by AtnNn** [comment]
There are lots of ways to do this, none of them seem ideal.

```
r.expr([{"a":1, "b":2}, {"a":3}, {"c":4}])
 .group(r.row.keys(),{multi:true}).ungroup()
 .map([r.row('group'), r.row('reduction')(r.row('group'))])
 .coerceTo('object')
```
